File: LeetCode/Medium/3202-Find_The_Maximum_Length_Of_Valid_Subsequence_II/3202-FindTheMaximumLengthOfValidSubsequenceII.cpp
#include <vector>

using namespace std;

class Solution
{
public:
    int maximumLength(vector<int> &nums, int k)
    {
        vector<vector<int>> dp(k, std::vector<int>(k, 0));
        int longest = 0;

        for (int num : nums)
        {
            int curr = num % k;

            for (int prev = 0; prev < k; prev++)
            {
                dp[curr][prev] = dp[prev][curr] + 1;
                longest = max(longest, dp[curr][prev]);
            }
        }

        return longest;
    }
};

File: LeetCode/Medium/3202-Find_The_Maximum_Length_Of_Valid_Subsequence_II/3202-FindTheMaximumLengthOfValidSubsequenceII.py
class Solution:
    def maximumLength(self, nums: list[int], k: int) -> int:
        dp = [[0] * k for _ in range(k)]
        longest = 0

        for num in nums:
            curr = num % k

            for prev in range(k):
                dp[curr][prev] = dp[prev][curr] + 1
                longest = max(longest, dp[curr][prev])

        return longest


a = Solution()
print(a.maximumLength(nums=[1, 2, 3, 4, 5], k=2))
print(a.maximumLength(nums=[1, 4, 2, 3, 1, 4], k=3))
